layout/plan.h

```diff
@@ -0,0 +1,428 @@
+/* -*- C -*- */
+/*
+ * Copyright (c) 2016-2021 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ * Original author: Nikita Danilov <nikita.danilov@seagate.com>
+ * Original creation date: 07-Jun-2016
+ */
+
+#pragma once
+
+#ifndef __MOTR_LAYOUT_PLAN_H__
+#define __MOTR_LAYOUT_PLAN_H__
+
+struct m0_layout {
+	struct m0_entity la_ent;
+};
+
+struct m0_composite_layout_layer {
+	uint32_t      cll_idx;
+	struct m0_fid cll_fid;
+};
+
+void m0_composite_layout_init(struct m0_composite_layout *layout,
+			      const struct m0_fid *fid);
+uint32_t m0_composite_layout_nr(const struct m0_composite_layout *layout);
+struct m0_composite_layout_layer *
+m0_composite_layout_at(const struct m0_composite_layout *layout, uint32_t idx);
+void m0_composite_layout_add(struct m0_composite_layout *layout, uint32_t idx);
+void m0_composite_layout_del(struct m0_composite_layout *layout, uint32_t idx);
+
+void m0_composite_layout_write(struct m0_composite_layout *layout,
+			       struct m0_dtx *dtx,
+			       struct m0_op **op);
+
+void m0_composite_layer_index(struct m0_composite_layout_layer *layer,
+			      bool write_mask, struct m0_idx *idx);
```

Comment:
@nikitadanilov what are these structs and functions in the beginning and how they are related to the rest of the API?

Reply:
@andriytk  here is something I remember when implementing composite layout (for Sage1 project): a composite object operation usually needs to issue a few ops,  such as querying extents from index and then sending IO ops, currently the composite object code does these ops in sequential steps, meaning we have to wait for (or block on) one step until we can move to next steps. Another is for the error handling in composite layout, when one op fails, how to deal with the whole IO.  Back then when implementing composite layout, layout plan was one of the option, but implementing it means the whole clovis IO code had to be re-written. So Sage1 didn't do it.

If we're going to implement the layout plan and use it for isc only, then motr may have 2 set of code doing similar thing (motr client IO code and layout plan), I am not sure if it is a good idea. But if motr is going to refactor client side code to use layout plan, that is going to be a huge job.

Reply:
According to @nikitadanilov these two structures and functions were added by mistake to this file, and they just can be removed.